src/test.cpp

```diff
@@ -0,0 +1,297 @@
+/**********************************************************************
+ * Copyright (c) 2018,2021 Pieter Wuille, Greg Maxwell, Gleb Naumenko *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file LICENSE or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#include "../include/minisketch.h"
+#include <vector>
+#include <algorithm>
+#include <limits>
+#include <random>
+#include <stdio.h>
+#include "util.h"
+
+namespace {
+
+uint64_t Combination(uint64_t n, uint64_t k) {
+    if (n - k < k) k = n - k;
+    uint64_t ret = 1;
+    for (uint64_t i = 1; i <= k; ++i) {
+        ret = (ret * n) / i;
+        --n;
+    }
+    return ret;
+}
+
+/** Create a vector with Minisketch objects, one for each implementation. */
+std::vector<Minisketch> CreateSketches(uint32_t bits, size_t capacity) {
+    if (!Minisketch::BitsSupported(bits)) return {};
+    std::vector<Minisketch> ret;
+    for (uint32_t impl = 0; impl <= Minisketch::MaxImplementation(); ++impl) {
+        if (Minisketch::ImplementationSupported(bits, impl)) {
+            ret.push_back(Minisketch(bits, impl, capacity));
+            CHECK((bool)ret.back());
+        } else {
+            CHECK(impl != 0); // implementation 0 must always work
+        }
+    }
+    return ret;
+}
+
+/** Test properties by exhaustively decoding all 2**(bits*capacity) sketches
+ *  with specified capacity and bits. */
+void TestExhaustive(uint32_t bits, size_t capacity) {
+    auto sketches = CreateSketches(bits, capacity);
+    if (sketches.empty()) return;
```

Comment:
perhaps assert that it's only empty when it's logically sensible for it to be empty? otherwise making empty always true bypasses this test.

Reply:
Done; just turned it into `CHECK(!sketches.empty())` for now. Can revisit when building support for certain bits field sizes is made optional.